src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1498 @@
+        if (r < 0)
+                log_error_errno(r, "Cannot emit JobRemoved message, ignoring: %m");
+
+        if (j->dbus_msg && j->complete_cb) {
+                if (sd_bus_error_is_set(&error))
+                        sd_bus_reply_method_error(j->dbus_msg, &error);
+                else {
+                        r = j->complete_cb(j->dbus_msg, j, json, &error);
+                        if (r < 0 || sd_bus_error_is_set(&error))
+                                sd_bus_reply_method_errno(j->dbus_msg, r, &error);
+                }
+        }
+
+        job_unref(j);
+        return 0;
+}
+
+static inline const char* sysupdate_binary_path(void) {
+        return secure_getenv("SYSTEMD_SYSUPDATE_PATH") ?: SYSTEMD_SYSUPDATE_PATH;
+}
+
+static int job_start(Job *j) {
+        _cleanup_close_ int stdout_fd = -EBADF;
+        _cleanup_(sigkill_waitp) pid_t pid = 0;
+        int r;
+
+        assert(j);
+
+        if (j->target->busy)
+                return -EBUSY;
+
+        stdout_fd = memfd_new("sysupdate-stdout");
+        if (stdout_fd < 0)
+                return stdout_fd;
+
+        r = safe_fork_full("(sd-sysupdate)",
+                           (int[]) { -EBADF, stdout_fd, STDERR_FILENO }, NULL, 0,
+                           FORK_RESET_SIGNALS|FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|
+                           FORK_REARRANGE_STDIO|FORK_LOG|FORK_REOPEN_LOG, &pid);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                /* Child */
+
+                _cleanup_free_ char *target_arg = NULL;
+                const char *cmd[] = {
+                        "systemd-sysupdate",
+                        "--json=short",
+                        NULL, /* maybe --component=, --root=, or --image= */
+                        NULL, /* maybe --offline */
+                        NULL, /* list, check-new, update, vacuum */
+                        NULL, /* maybe version (for list, update) */
+                        NULL
+                };
+                unsigned k = 1;
+
+                if (setenv("NOTIFY_SOCKET", "/run/systemd/sysupdate/notify", 1) < 0) {
+                        log_error_errno(errno, "setenv() failed: %m");
+                        _exit(EXIT_FAILURE);
+                }
+
+                r = setenv_systemd_exec_pid(true);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to update $SYSTEMD_EXEC_PID, ignoring: %m");
+
+                if (j->target->class == TARGET_HOST)
+                        r = 0; /* Host doesn't need any argument */
+                else if (j->target->class == TARGET_COMPONENT)
+                        r = asprintf(&target_arg, "--component=%s", j->target->name);
+                else if (IN_SET(j->target->image_type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME))
+                        r = asprintf(&target_arg, "--root=%s", j->target->path);
+                else
+                        r = asprintf(&target_arg, "--image=%s", j->target->path);
```

Comment:
please use strjoin() for concatenating multiple strings. vastly faster (and shorter) than asprintf(). It's fine to use asprintf() for more complex formatting, but not for such trivial stuff.